### PR TITLE
Fix #2013: separate interpreter class from its results

### LINKED ISSUE #2013: Separate `Interpreter` class from its results
The results of the `Interpreter` class running over Python bytecode could be obtained through other ways, such as loop-lifting (see #1996). We should therefore probably separate those two concepts into separate classes.

### PR DESCRIPTION
**Pull request by pitrou**
This necessitates other pieces of refactoring to produce a clean outcome. As a result, the `Interpreter` class is further split: the `Interpreter` class is now only responsible for iterating over bytecode and producing a basic Numba IR from it. A separate `PostProcessor` class handles canonicalization and analysis of the IR (computing variable lifetimes, etc.). The IR is represented as a `FunctionIR` object.

There's also an additional `FunctionIdentity` class that supersedes `FunctionAttributes`. I expect that some parts of `FunctionDescriptor` should/will be moved to it (the parts that are simple function metadata rather than the outcome of parts of the Numba compilation pipeline).

Also we see that `FunctionIR` is able to entirely represent the outcome of the Numba "front-end". This should make us able to split the pipeline into three parts: "front-end", "middle-end" (rewrites + typeinfer), "back-end" (lowering).

### CODE REVIEW

File: numba/tests/test_loopdetection.py [hunk 1]
```diff
@@ -1,47 +0,0 @@
-from __future__ import print_function
```
Comment (pitrou) [inline numba/tests/test_loopdetection.py]: Note these tests were superseded a long time ago by the CFG tests in `test_flow_control.py`.

File: numba/ir.py [hunk 2]
```diff
@@ -782,5 +795,111 @@ def __repr__(self):
         return "Loop(entry=%s, exit=%s)" % args
 
 
+class FunctionIR(object):
+
+    def __init__(self, interp):
+        # XXX turn this inside out?  Interpreter should first construct
+        # a FunctionIR object and then work on it
```
Comment (sklam) [inline numba/ir.py]: Agree.  Preferably, FunctionIR should not reference interp.

File: numba/funcdesc.py [hunk 3]
```diff
@@ -131,19 +131,20 @@ def __repr__(self):
...
         -------
         qualname, unique_name, modname, doc, args, kws, globals
 
         ``unique_name`` must be a unique name.
         """
-        func = interp.bytecode.func
-        qualname = interp.bytecode.func_qualname
+        func = func_ir.func_id.func
+        qualname = func_ir.func_id.func_qualname
+        # XXX to func_id
```
Comment (sklam) [inline numba/funcdesc.py]: wouldn't it be a simple change that can be done now?

File: numba/interpreter.py [hunk 4]
```diff
@@ -60,298 +60,58 @@ def get_assignment_source(self, destname):
...
-                        index = len(dct) + 1
-                        yieldinst.index = index
-                        yp = YieldPoint(block, yieldinst)
-                        dct[yieldinst.index] = yp
-
-    def _compute_generator_info(self):
-        """
-        Compute the generator's state variables as the union of live variables
-        at all yield points.
+    def result(self):
```
Comment (sklam) [inline numba/interpreter.py]: Empty docstring?

File: numba/compiler.py [hunk 5]
```diff
@@ -179,6 +142,22 @@ def compile_isolated(func, args, return_type=None, flags=DEFAULT_FLAGS,
                              flags, locals)
 
 
+def run_frontend(func):
+    """
+    Run the compiler frontend over the given Python function, and return
+    the function's canonical Numba IR.
+    """
+    # XXX make this a dedicated Pipeline?
```
Comment (sklam) [inline numba/compiler.py]: Since this is only used by tests, can we make it reuse the code in the pipeline?

File: numba/funcdesc.py (hunk 3 above)
Comment (sklam) [inline numba/funcdesc.py]: this function can just use `func_id` instead of `func_ir`.

Comment (sklam) [inline numba/funcdesc.py]: > this function can just use func_id instead of func_ir.

nvm, due to `func_ir.arg_names` vs `func_id.arg_names`

Comment (pitrou) [inline numba/funcdesc.py]: I've experimented with this, and it's not that easy. With `generated_jit`, there is a distinction between the function being implemented (the decorated top-level function) and the implementation being compiled. The `func_id` denotes the implementation being compiled...

File: numba/compiler.py (hunk 5 above)
Comment (pitrou) [inline numba/compiler.py]: As the comment suggests, I think we should make it a separate pipeline (i.e. split the current unified pipeline in three sub-pipelines). Do you think that makes sense?

Comment (sklam) [inline numba/compiler.py]: what are the three?  frontend, typeinfer, backend?

Comment (pitrou) [inline numba/compiler.py]: Yes. I would also group the two rewrites passes with typeinfer.

Comment (sklam) [inline numba/compiler.py]: go for it =)

**Pull request merged**